### PR TITLE
CAN Support

### PR DESCRIPTION
**Pull request by zklapow**
This PR is fairly incomplete still, though it is functional.

I created a (temp) create called `embedded-hal-can` to try implementing the traits currently in https://github.com/rust-embedded/embedded-hal/pull/77 though that PR is so old now, I am not sure its going to get much traction (and I am not totally sure I love the structure of those traits anyhow) so I may remove that piece.

I have been testing this on an `STM32f302C8T6`, I am working on simplifying my "blinky with can" into an example of how to use this.

todos:
- [x] Generify this to support other stm32f3xx devices

future improvements:
- Some sort of baud rate param heuristics (maybe just embed a table for standard rates?)
- Improved filter handling (the current implementation "wastes" filter registers)

Wanted to post this now though as I have been sitting on it for a while and I figured it would be good to at least get thoughts/feedback.

### CODE REVIEW

File: src/can.rs
Comment (Sh3Rm4n) [inline src/can.rs:11]: Has to be adjusted to use the controller declared by the feature flag.

File: src/can.rs
Comment (Sh3Rm4n) [inline src/can.rs:22]: Is `#derive(Hash)` needed for anything?

File: Cargo.toml
Comment (Sh3Rm4n) [inline Cargo.toml:27]: I don't see were `typenum` is used.

File: Cargo.toml
Comment (Sh3Rm4n) [inline Cargo.toml:28]: Hm. As this is an experimental crate until this is added to the `embedded-hal` we should feature gate the CAN implementation as a whole?

But I'm happy to include it into this hal, so that more people can experiment with the can abstraction! :)

File: src/can.rs
Comment (Sh3Rm4n) [inline src/can.rs:192]: Even though rust has the `new` convention for constructors, naming the constructor as the peripheral name is really common in the rust hal implementations. This is import e.g. for SPI to differentiate between `spi1, spi2, ...`. As stm32f3 devices do not have multiple can peripherals this is not as important, but maybe we should stick to it and name it `can`?

File: src/can.rs
Comment (Sh3Rm4n) [inline src/can.rs:197]: The can peripheral is not well documented atm as of [stm32-rs](https://stm32-rs.github.io/stm32-rs/STM32F302.html#CAN), but this could be named

```suggestion
        apb1.enr().modify(|_, w| w.canen().enabled());
```

[doc link](https://docs.rs/stm32f3/0.11.0/stm32f3/stm32f303/rcc/apb1enr/struct.CANEN_W.html#method.enabled)

Maybe you could try to apply these to methods, where it is possible right now :)

File: src/can.rs
Comment (Sh3Rm4n) [inline src/can.rs:299]: Hm. I think  we should generally thrive towards not using asserts inside this hal. (There are enough other places in the hal, where we do use asserts) But  this assert is unfortunate, as we need the `nb::Error` to return. 

Do you have any ideas?

File: src/can.rs
Comment (Sh3Rm4n) [inline src/can.rs:330]: `if let Some(..)` could be used here, I guess :thinking:

File: src/can.rs
Comment (Sh3Rm4n) [inline src/can.rs:357]: I lost the meaning of `i` here. Maybe a more descriptive name is needed.

File: src/can.rs
Comment (Sh3Rm4n) [inline src/can.rs:487]: Just curious. Why must the data slice be converted into an array? Slice has a [`len()`](https://doc.rust-lang.org/std/primitive.slice.html#method.len) method. Maybe you could give me some insight :)

File: src/can.rs
Comment (Sh3Rm4n) [reply src/can.rs:299]: While reading through the code, isn't this assert already handled in `CanFrame`?

File: src/can.rs
Comment (zklapow) [reply src/can.rs:487]: I should likely limit the visibility of this method and it could be simplified a bit... The reason this exists is mainly to do with lifetimes and there are sort of two reasons for it:

1. is for reading, we allocate an array always of length 8 bytes (the max CAN frame size), because we cant do dynamic allocation and then use a subslice of it, but now that I think more about it, we can likely just consume that array in this case and avoid the reference.
2. Is because of how we need to implement `embedded_hal_can::Frame`. Having every frame have a 8 byte backing array was just the simplest way to handle this without getting into using `generic_array` or something, but ultimately I may do that, because it would make the code clearer.

File: src/can.rs
Comment (zklapow) [reply src/can.rs:299]: ya ill remove this

File: src/can.rs
Comment (zklapow) [reply src/can.rs:197]: 👍 a bunch of this was written before some upgrades to the svd-to-rust stuff which I suspect is the source of this.

File: Cargo.toml
Comment (zklapow) [reply Cargo.toml:28]: ya my plan was to put this whole thing behind a feature flag once it was more merge-ready, but intellij doesn't totally know how to handle feature flags ATM so I have avoided that for ease of development.

**Comment by zklapow** [comment]
Added a simple "blinky" example, heres a GIF of it in action: 

Image

**Comment by zklapow** [comment]
@Sh3Rm4n this should be ready for another look when you have some time, it should be mergeable in its current state.

A couple notes:

- I updated to use `ArrayVec` to give the `CanFrame` a slightly simpler API (no need to remember that there is an extra `length` field there!). I tried to use `generic_array` but I don't think that will be possible without having `transmit` be generic of the frame size. Given that a frame is at most 8 bytes, having to allocate the full backing array seemed OK. (I also considered introducing some sort of allocator dedicated to canframes but.... that seemed extreme)
- Everything should be behind the `can` feature flag now, so it is disabled by default.

**Comment by zklapow** [comment]
@Sh3Rm4n any thoughts on another review or merging this?

**Comment by zklapow** [comment]
I moved those around a bit to be more clear, I do think they would be interesting to tackle in the future but are sort of hard and don't really affect the functionality here much. This should be useable and mergeable as-is.

File: Cargo.toml
Comment (Sh3Rm4n) [inline Cargo.toml:35]: [heapless](https://crates.io/crates/heapless) is an alternative and more widely used in the embedded community, FYI. 

But I do not care and haven't used both enough to have an opion, if one or the other would be better for this use case.

File: Cargo.toml
Comment (Sh3Rm4n) [inline Cargo.toml:108]: Should be reverted I guess :)

File: Cargo.toml
Comment (Sh3Rm4n) [inline Cargo.toml:120]: As the [example](https://github.com/stm32-rs/stm32f3xx-hal/pull/100/files#diff-f005bfff7df113692409278f2a71eec9R7) explicitly uses the `stm32f303` we should requre the "stm32f303" feature instead of the "device-sected" feature: 

```suggestion
required-features = ["rt", "can", "stm32f303"]
```

"device-seletected" feature is not really meant to be used outside the crate and is an implementation detail.

File: examples/can.rs
Comment (Sh3Rm4n) [inline examples/can.rs:60]: Nice, this is also now an example for the watchdog :+1:

File: examples/can.rs
Comment (Sh3Rm4n) [inline examples/can.rs:42]: AF7 or AF9?

File: src/can.rs
Comment (Sh3Rm4n) [inline src/can.rs:9]: Could also be written as

```suggestion
use nb::{self, Error};
```

File: src/can.rs
Comment (Sh3Rm4n) [inline src/can.rs:14]: Would be more readable this way:

```suggestion
const EXID_MASK: u32 = 0b11111_11111100_00000000_00000000;
const MAX_EXTENDED_ID: u32 = 0x1FFF_FFFF;
```

File: src/can.rs
Comment (Sh3Rm4n) [inline src/can.rs:51]: What the meaning / purpose of these underscores?

File: src/can.rs
Comment (Sh3Rm4n) [inline src/can.rs:49]: I would like to see more comments / documentation, especially on the public interfaces, like this struct :)

File: src/can.rs
Comment (Sh3Rm4n) [inline src/can.rs:322]: A comment on this `unsafe` instance would be great. And couldn't this `unsafe` scope be limited insid3 the `math j { ... }`?

File: src/can.rs
Comment (Sh3Rm4n) [inline src/can.rs:358]: I would make this explicit and not hide it behind a function, as it is done everywhere else in this hal IIRC

```suggestion
        let can = unsafe { &*stm32::CAN::ptr() };
```

File: src/can.rs
Comment (zklapow) [reply src/can.rs:322]: I'll add a note, there are actually 2 unsafe calls here which is the reason for the outer scope (the tdtr call below also calls the unsafe `bits`) but these can be split. The note here will probably be somewhat silly, I'd have to do some more digging but I wonder if we could get svd-to-rust to generate safe calls for these registers, as the only actual way to update them is via a full register write

File: src/can.rs
Comment (zklapow) [reply src/can.rs:50]: this just indicates that these fields are unused

**Comment by zklapow** [comment]
I have updated this to address the open comments. I added a bunch of notes about the usages of unsafe but a lot of those notes are semi silly but calling `bits()` is just always unsafe but also the only way to access those registers.

**Comment by David-OConnor** [comment]
Merge!

**Comment by zklapow** [comment]
@David-OConnor would love to tho I dont have access 👍 

@Sh3Rm4n or @ra-kete I have merged master to address some conflicts as well as updated the changelog, and the build is passing. Any chance one of you could merge this for me when you have a moment? I think all of the comments should be addressed.

**Comment by ra-kete** [comment]
@Sh3Rm4n has reviewed this, so if he gives his thumbs-up for merging I'm fine too.

File: src/can.rs
Comment (Sh3Rm4n) [inline src/can.rs:286]:
```suggestion
    pub fn free(self) -> stm32::CAN {
```

And it should release `gpioa::PA11<AF9>` and    `gpioa::PA12<AF9>` as well.

**Review by Sh3Rm4n — APPROVED** [review]
Hey, sorry for the late response. 

Minor nits and [this](https://github.com/stm32-rs/stm32f3xx-hal/pull/100/files#r456956829) is not addressed, yet. 

But overall LGTM. :)

**Comment by zklapow** [comment]
updated, the docs are still a bit sparse but unfortunately I don't have a ton of time to work on this ATM.

**Comment by Sh3Rm4n** [comment]
Nice. Thanks for the contribution. Finally i've time to merge this 🎉 

Thanks for your patience. Further things can still be improved later on, if necessary.

**Pull request merged**
